### ISSUE TITLE
Random bits (SRFI 27) support

### ISSUE THREAD

**Issue by jasonmacduffie** [justinethier/cyclone#70]
Hello, are there plans to add built-in support for random bits?

http://srfi.schemers.org/srfi-27/srfi-27.html

**Comment by justinethier** [comment]
@Taknamay Sure, this would be a great addition and we need support for random numbers at some point. 

It might be a little while before I get to it, though. Are you interested in starting the work on any of this?

**Comment by Taknamay** [comment]
Hmm, it would be the first time I wrote something like this. I could give it a try.

**Comment by Taknamay** [comment]
Ok, I think I have made progress on this. However I am a bit confused on how the FFI works.

For example, this compiles OK:
     7(define-c DONOTHINGFUNCTION
     8  "(void *data, int argc, closure _, object k, object z)"
     9  " return_inexact_double_op(data, k, log, z);")
    10
    11;; Numbers taken from bsd random                                              
    12(define-c next-lcg
    13  "(void *data, int argc, closure _, object k, object seed)"
    14  "return_exact_double_op(data,k,next_lcg,seed)")

But this does not:
    11;; Numbers taken from bsd random                                              
    12(define-c next-lcg
    13  "(void *data, int argc, closure _, object k, object seed)"
    14  "return_exact_double_op(data,k,next_lcg,seed)")

Also, how do I get the value of an object? For example, a Scheme integer into a C integer?

**Comment by justinethier** [comment]
First of all, thanks for taking the time to work on this.

Hmm... I'm not sure why the same snippet of code would not compile in the second case. You are missing a semicolon at the end of the last line, for what it's worth. What compile error(s) are you getting?

Unfortunately the FFI probably needs better documentation. Sorry about that.

There are a series of macros to convert between Scheme and C integers:
https://github.com/justinethier/cyclone/blob/master/include/cyclone/types.h#L216

Basically, if you are working with a value type (integers and characters) you need to use the appropriate macro to convert from Scheme to C. Otherwise, if you have an object type you need to cast to the appropriate C type (EG: [`pair_type`](https://github.com/justinethier/cyclone/blob/master/include/cyclone/types.h#L431) or [`double_type`](https://github.com/justinethier/cyclone/blob/master/include/cyclone/types.h#L326)) and then access structure members directly.

Does that help at all?

**Comment by Taknamay** [comment]
Yes, that helps a lot, thanks! And the semicolon was the problem after all. Weird that adding that code make the compiler accept it.

**Comment by justinethier** [comment]
Very weird. Anyway, glad that fixed the problem :)

**Comment by Taknamay** [comment]
Great, I got the basic generator working! By the way, are you ok with an [LCG](https://en.wikipedia.org/wiki/Linear_congruential_generator), at least for now?

**Comment by justinethier** [comment]
LCG seems like a good starting point but on the other hand the SRFI discourages it:

> The most important decision about the implementation is the choice of the random number generator. The basic principle here is: Let quality prevail! In the end, a performance penalty of a better generator may be a cheap price to pay for some avoided catastrophes. It may be unexpected, but I have also seen many examples where the better generator was also the faster. <b>Simple linear congruential generator cannot be recommended as they tend to be ill-behaved in several ways.</b>
For this reason, my initial proposal was George Marsaglia's COMBO generator, which is the combination of a 32-bit multiplicative lagged Fibonacci-generator with a 16-bit multiply with carry generator. The COMBO generator passes all tests of Marsaglia's DIEHARD testsuite for random number generators and has a period of order 2^60.

> As an improvement, Brad Lucier suggested suggested Pierre L'Ecuyer's MRG32k3a generator which is combination of two recursive generators of degree three, both of which fit into 54-bit arithmetics. The MRG32k3a generator also passes DIEHARD and in addition, has desireable spectral properties and a period in the order of 2^191. As a matter of fact, multiple recursive generators (MRGs) are theoretically much better understood than special constructions as the COMBO generator. This is the reason why the implementations provided here implements the MRG32k3a generator. When implemented in floating point arithmetics with sufficient mantissa-width, this generator is also very fast.

I don't know anything about MRG32k3a. A quick google search finds this:
http://simul.iro.umontreal.ca/rng/MRG32k3a.c

Though there may be more to it:
http://srfi.schemers.org/srfi-27/mail-archive/msg00023.html

**Comment by jasonmacduffie** [comment]
Yes, I was thinking the LCG would just help to get something out the door. Definitely something higher quality would be good eventually.

**Comment by justinethier** [comment]
This is a good implementation for now, and maybe it can be enhanced further in the future.
